rdr_service/tools/tool_libs/pdr_update_tool.py

```diff
@@ -0,0 +1,174 @@
+        """
+        try:
+            rebuild_bq_participant(pid, ps_bqgen=ps_bqgen, pdr_bqgen=pdr_bqgen, project_id=self.gcp_env.project)
+        except NotFound:
+            return 1
+        return 0
+
+    def update_many_pids(self, pids):
+        """
+        Update many pids from a file.
+        :return:
+        """
+        if not pids:
+            return 1
+
+        ps_bqgen = BQParticipantSummaryGenerator()
+        pdr_bqgen = BQPDRParticipantSummaryGenerator()
+
+        total_pids = len(pids)
+        count = 0
+        errors = 0
+
+        for pid in pids:
+            count += 1
+
+            if self.update_single_pid(pid, ps_bqgen=ps_bqgen, pdr_bqgen=pdr_bqgen) != 0:
+                errors += 1
+                if self.args.debug:
+                    _logger.error(f'PID {pid} not found.')
+
+            if not self.args.debug:
+                print_progress_bar(
+                    count, total_pids, prefix="{0}/{1}:".format(count, total_pids), suffix="complete"
+                )
+
+        if errors > 0:
+            _logger.warning(f'\n\nThere were {errors} PIDs not found during processing.')
+
+        return 0
+
+
+    def run(self):
+        """
+        Main program process
+        :return: Exit code value
+        """
+        clr = self.gcp_env.terminal_colors
+        pids = None
+
+        if not self.args.from_file and not self.args.pid:
+            _logger.error('Nothing to do')
+            return 1
+
+        _logger.info(clr.fmt('\nRebuild Participant Summaries for PDR:', clr.custom_fg_color(156)))
+        _logger.info('')
+        _logger.info('=' * 90)
+        _logger.info('  Target Project        : {0}'.format(clr.fmt(self.gcp_env.project)))
+
+        if self.args.from_file:
+            filename = os.path.expanduser(self.args.from_file)
+            if not os.path.exists(filename):
+                _logger.error(f"File '{self.args.from_file}' not found.")
+                return 1
+
+            # read pids from file.
+            pids = open(os.path.expanduser('~/rebuild_pids.txt')).readlines()
+            # convert pids from a list of strings to a list of integers.
+            pids = [int(i) for i in pids]
+            _logger.info('  PIDs File             : {0}'.format(clr.fmt(self.args.from_file)))
+            _logger.info('  Total PIDs            : {0}'.format(clr.fmt(len(pids))))
+        else:
+            _logger.info('  PID                   : {0}'.format(clr.fmt(self.args.pid)))
+
+        _logger.info('=' * 90)
+        _logger.info('')
+
+        self.gcp_env.activate_sql_proxy()
+        _logger.info('')
+
+        if self.args.from_file:
+            return self.update_many_pids(pids)
+
+        if self.args.pid:
+            if self.update_single_pid(self.args.pid) == 0:
+                _logger.info(f'Participant {self.args.pid} updated.')
+            else:
+                _logger.error(f'Participant ID {self.args.pid} not found.')
+
+        return 1
+
+
+def run():
```

Comment:
I think having a run method and a run function could be potentially confusing. Perhaps we could call this `main`

Reply:
If you want to change it, I don't mind.  We should probably do it under a different ticket and change it in all the new tools.

Reply:
Ok, we’ll call that a todo 👍